File: algorithm/235.cpp
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root==nullptr || root==p || root==q) return root;

        if((root->val>p->val && root->val<q->val) || (root->val<p->val && root->val>q->val)) //在root左右子树上，直接返回root
        {
            return root;
        }
        else if(root->val>p->val && root->val>q->val)  //在root左子树上，找到第一个目标节点，它就是两个节点的最近公共祖先
        {
            return lowestCommonAncestor(root->left, p, q);
        }
        else return lowestCommonAncestor(root->right, p, q);  //在root右子树上，找到第一个目标节点，它就是两个节点的最近公共祖先
    }
};

File: algorithm/go/firstround/235.go
package firstround

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 时间复杂度O(n),空间复杂度O(1)
func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	var minVal, maxVal int
	if p.Val <= q.Val {
		minVal = p.Val
		maxVal = q.Val
	} else {
		minVal = q.Val
		maxVal = p.Val
	}

	// p,q分别在root的左右子树上,或者p,q中的一个节点为当前root节点
	if (root.Val > minVal && root.Val < maxVal) || root == p || root == q {
		return root
	} else if root.Val > maxVal { // root.Val比两个节点值都大,往左子树查找
		return lowestCommonAncestor(root.Left, p, q)
	} else if root.Val < minVal { // root.Val比两个节点值都小,往右子树查找
		return lowestCommonAncestor(root.Right, p, q)
	}

	return nil
}
